### DISCUSSION TITLE
Unreal : GetAGLevel returns NaN so the airplane crashes instantly

### DISCUSSION THREAD

**Discussion by demonixis** [JSBSim-Team/jsbsim#862] [Support requests]
Hello,
I'm integrating JSBSim into my simulator. I've an issue after the takeoff, the airplane is teleported instantly far away few seconds after the takeoff. It's perfect at the begining then it's teleported for some reason.

I've a GeoReference actor on the scene, it's setup the same than the Unreal demo project into this repository. I got this error

``` 
LogOutputDevice: Error: === Handled ensure: ===
LogOutputDevice: Error: Ensure condition failed: !Result->ContainsNaN()  [File:D:\build\++UE5\Sync\Engine\Source\Runtime\Core\Private\Math\UnrealMathSSE.cpp] [Line: 165] 
LogOutputDevice: Error: Stack: 
LogOutputDevice: Error: [Callstack] 0x00007ffd19f4074d UnrealEditor-Core.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffcd1975c0e UnrealEditor-GeoReferencing.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffcd0069b61 UnrealEditor-JSBSimFlightDynamicsModel-0010.dll!UJSBSimMovementComponent::GetAGLevel() [E:\Projects\Unreal\MiniFS-UE4\Plugins\JSBSimFlightDynamicsModel\Source\JSBSimFlightDynamicsModel\Private\JSBSimMovementComponent.cpp:191]
LogOutputDevice: Error: [Callstack] 0x00007ffcd00699c3 UnrealEditor-JSBSimFlightDynamicsModel-0010.dll!UEGroundCallback::GetAGLevel() [E:\Projects\Unreal\MiniFS-UE4\Plugins\JSBSimFlightDynamicsModel\Source\JSBSimFlightDynamicsModel\Private\UEGroundCallback.cpp:24]
LogOutputDevice: Error: [Callstack] 0x0000019552d63815 JSBSim.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x0000019552d60847 JSBSim.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x0000019552d9af7f JSBSim.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x0000019552d97dd0 JSBSim.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x0000019552d95df0 JSBSim.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffcd0068729 UnrealEditor-JSBSimFlightDynamicsModel-0010.dll!UJSBSimMovementComponent::DoTrim() [E:\Projects\Unreal\MiniFS-UE4\Plugins\JSBSimFlightDynamicsModel\Source\JSBSimFlightDynamicsModel\Private\JSBSimMovementComponent.cpp:708]
LogOutputDevice: Error: [Callstack] 0x00007ffcd006df43 UnrealEditor-JSBSimFlightDynamicsModel-0010.dll!UJSBSimMovementComponent::PrepareJSBSim() [E:\Projects\Unreal\MiniFS-UE4\Plugins\JSBSimFlightDynamicsModel\Source\JSBSimFlightDynamicsModel\Private\JSBSimMovementComponent.cpp:543]
LogOutputDevice: Error: [Callstack] 0x00007ffcd006e063 UnrealEditor-JSBSimFlightDynamicsModel-0010.dll!UJSBSimMovementComponent::PrepareModelForCompVisualizer() [E:\Projects\Unreal\MiniFS-UE4\Plugins\JSBSimFlightDynamicsModel\Source\JSBSimFlightDynamicsModel\Private\JSBSimMovementComponent.cpp:1197]
LogOutputDevice: Error: [Callstack] 0x00007ffcc3251fb2 UnrealEditor-JSBSimFlightDynamicsModelEditor-0010.dll!FJSBSimMovementCompVisualizer::DrawVisualization() [E:\Projects\Unreal\MiniFS-UE4\Plugins\JSBSimFlightDynamicsModel\Source\JSBSimFlightDynamicsModelEditor\Private\JSBSimMovementCompVisualizer.cpp:54]
LogOutputDevice: Error: [Callstack] 0x00007ffd14ba0a1a UnrealEditor-UnrealEd.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd1469400b UnrealEditor-UnrealEd.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd094029c7 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd088b2f1f UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd09420f2b UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd0941c013 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd0941c9a0 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd14202545 UnrealEditor-UnrealEd.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd0fff9758 UnrealEditor-Engine.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd14183d24 UnrealEditor-UnrealEd.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd1417c665 UnrealEditor-UnrealEd.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd14bfe106 UnrealEditor-UnrealEd.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ff710d0852c UnrealEditor.exe!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ff710d2287d UnrealEditor.exe!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ff710d2296a UnrealEditor.exe!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ff710d25680 UnrealEditor.exe!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ff710d371b4 UnrealEditor.exe!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ff710d3a516 UnrealEditor.exe!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd866726bd KERNEL32.DLL!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007ffd8772a9f8 ntdll.dll!UnknownFunction []
```

Any clues?

Edit: The GetAGLevel function generates a NAN value, which break the function call. After that the aircraft status is crashed.

Looks like it's the same problem in [this issue](https://github.com/JSBSim-Team/jsbsim/issues/786)

Thanks

**Comment by gallonmate** [comment]
I know you already started an issue but I figured I would copy/paste my reply there for others in the future:

> I would start again with UE5 reference demo project and slowly introduce whatever changes you added to help find why the crash happens: The aircraft used, the location, the meshes used for ground, geo-reference settings, etc.
> 
> Until you or others can reproduce the issue using the Reference project demo level and demo airplane blueprint, then there's not much that can be done. JSBSim currently allows for many settings to be initialized that can easily produce NaNs and JSBSim will not crash. It's UE5 that crashes from the NaNs. The issue in finding out what is causing the NaN in the first place. Which I why I suggest starting with the default demo project and adding changes to it, until NaN happens again.

**Comment by agodemar** [comment]
@demonixis those logs, without a detailed explanation of what you actually did with your code, are far from being a clue to the reader. If you want to get some help I think you should give more context: What's your FDM model, what are the initial conditions, what sequence of inputs are you providing to the model, etc.

**Comment by demonixis** [comment]
@agodemar here is what I've discovered so far...

## Setup
Unreal: 5.1.1
Model: Cessna c172p
Initial conditions:
- Start on ground: False or True 
- Start with Gear Down (True)
- Start With Engine Running (True)
- Flaps 0
- Calibrated Air Speed 0
- Default Atmosphere settings

## App Setup and Pawn
The aircraft is not possessed by the Player. Instead a Character is parented to the Airplane, and this is the character that controls the airplane. It allows me to have two pilots controlling the aircraft at the same time. Collisions are disabled on Character before parenting them to the Aircraft.

Collisions on the Aircraft are disabled too because it's handled by GroundCheck. Simulate Physics is also disabled, exactly like the Unreal example.

## Update Code
I've an `Aircraft` class that inherits from `APawn`. This class contains a method `UpdateJSBSim` with the following code called in the `Tick` method.

```cpp
void AAircraft::UpdateJSBSim(const float DeltaTime)
{
#if PLATFORM_WINDOWS
	UJSBSimMovementComponent* JsbSimFdm = Cast<UJSBSimMovementComponent>(JsbSimComponent);

	if (JsbSimFdm == nullptr) return;
	
	// Update Commands
	FFlightControlCommands& Commands = JsbSimFdm->Commands;
	Commands.Aileron = RollInput;
	Commands.Elevator = PitchInput;
	Commands.Rudder = YawInput;
	Commands.Flap = FlapsInput;
	Commands.Spoiler = 0;
	Commands.Steer = YawInput;
	Commands.CenterBrake = BrakesInput;
	Commands.LeftBrake = BrakesInput;
	Commands.GearDown = 1;
	Commands.ParkingBrake = ParkingBrakesInput;
	Commands.PitchTrim = PitchTrim;
	Commands.RollTrim = RollTrim;
	Commands.RightBrake = BrakesInput;
	Commands.SpeedBrake = 0;
	Commands.YawTrim = 0;
	
	for(int i = 0; i < JsbSimFdm->EngineCommands.Num(); i++)
	{
		FEngineCommand& Engine = JsbSimFdm->EngineCommands[i];
		Engine.Throttle = ThrottleInput;
		Engine.Magnetos = EMagnetosMode::Both;// static_cast<EMagnetosMode>(MagnetoInput);
		//Engine.Mixture = 0;// MixtureInput;
		Engine.Starter = bEngineStarted;// EngineStarterInput;
		Engine.Running = bEngineStarted;
	}
#endif
}
```
The values passed to Commands/Engines structures are clamped to -1/1 when they're float type.

## Behaviour
I can takeoff and sometimes fly. But I often get NaN and an exception in Unreal. I know how to reproduce the exception easily. For that I've to :
- Take some speed
- Take off
- Get some altitude ~100ft for instance
- Push the elevator stick (for instance to simulate an engine failure)

The airplane start to fall forward and get speed, then I got a lot of NaNs logged in the console with an exception. If the angle is low, I can land without problem. If the angle of attack is to high it gets NaNs.

## The logs
I added a lot of logs in the Unreal plugin, it seems that the GetAGLevel function causes the problem. I log the return value at each frame. it's interesting to note that this function returns 0 many time just before it gets NaN.

I've noted that data are sometime inconsistant, for instance the following data was captured during the airplane fall
```bash
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=1.353263
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=1.353263
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=-0.078834
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=-0.055326
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=-0.049509
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.551602
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=1.902418
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=1.931474
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=1.353263
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=-0.078834
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=-0.055326
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=-0.049509
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.551602
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=1.902418
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=1.931474
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=1.353278
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=1.353278
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=1.353278
```

The NaNs

```bash
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=53.533254
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=54.419378
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=52.996474
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=54.472643
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=58.711977
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=51.604072
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogNiagara: Warning: NiagaraComponent(NiagaraComponent /Game/Maps/99_Sandbox/UEDPIE_0_Airport_3.Airport_3:PersistentLevel.BP_Player_C_0.TeleportTraceNiagaraSystem - NiagaraSystem /Game/Vendors/VRTemplate/VFX/NS_TeleportTrace.NS_TeleportTrace) required LWC tile recache and was reset.
LogParticles: Warning: PSC(ParticleSystemComponent /Game/Maps/99_Sandbox/UEDPIE_0_Airport_3.Airport_3:PersistentLevel.BP_LaserPointer_GEN_VARIABLE_BP_LaserPointer_C_CAT_0.PS_LaserBeam - ParticleSystem /Game/Effects/LaserBeam/PS_LaserBeam.PS_LaserBeam) required LWC tile recache and was reset.
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogNiagara: Warning: NiagaraComponent(NiagaraComponent /Game/Maps/99_Sandbox/UEDPIE_0_Airport_3.Airport_3:PersistentLevel.BP_Player_C_0.TeleportTraceNiagaraSystem - NiagaraSystem /Game/Vendors/VRTemplate/VFX/NS_TeleportTrace.NS_TeleportTrace) required LWC tile recache and was reset.
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Error: UEGroundCallback::GetAGLevel FGLocation contains NaN
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Error: UEGroundCallback::GetAGLevel FGLocation contains NaN
```

The `FGLocation` variable comes from JSBSim with a NaN value. That means the Unreal Engine plugin is sending a bad value OR there is an error during calculation.

```cpp
double UEGroundCallback::GetAGLevel(const FGLocation& location, FGLocation& contact, FGColumnVector3& normal, FGColumnVector3& vel, FGColumnVector3& angularVel) const
{
	vel.InitMatrix();
	angularVel.InitMatrix();

	FVector ECEFLocation = FVector(location(1), location(2), location(3)) * FEET_TO_METER;

	if (ECEFLocation.ContainsNaN())
	{
		UE_LOG(LogJSBSim, Error, TEXT("UEGroundCallback::GetAGLevel FGLocation contains NaN"));
        // End of the code
}
```

After few frames a new NaN arrives

```bash
LogJSBSim: Display: UJSBSimMovementComponent::GetAGLevel HAT=0.000000
LogJSBSim: Error: UJSBSimMovementComponent::Tick EngineLocation contains NaN
```

In the following code (in the UJSBSimMovementComponent::Tick), we can see that `AircraftState.ECEFLocation` contains NaN. That's why other calculations are not good. This variable is set in the `CopyFromJSBSim` method. First NaN can be get here too.

```cpp
FTransform ENUTransform = GeoReferencingSystem->GetTangentTransformAtECEFLocation(
					AircraftState.ECEFLocation);
FRotator LocalUERotation(AircraftState.LocalEulerAngles);
LocalUERotation.Yaw = LocalUERotation.Yaw - 90.0;
// JSBSim heading is aero heading (0 at north). We have to remove 90 because in UE, 0 is pointing east. 
FQuat EngineRotationQuat = ENUTransform.TransformRotation(LocalUERotation.Quaternion());

FMatrix EngineRotation;
EngineRotationQuat.ToMatrix(EngineRotation);
FVector CGOffsetWorld = EngineRotation.TransformPosition(CGLocalPosition);

// Computes Location in engine frame
FVector CGWorldPosition;
GeoReferencingSystem->ECEFToEngine(AircraftState.ECEFLocation, CGWorldPosition);
FVector EngineLocation = CGWorldPosition - CGOffsetWorld;
```

And the the exception

```bash
LogOutputDevice: Error: === Handled ensure: ===
LogOutputDevice: Error: Ensure condition failed: Y != 0  [File:D:\build\++UE5\Sync\Engine\Source\Runtime\Core\Private\GenericPlatform\GenericPlatformMath.cpp] [Line: 124] 
LogOutputDevice: Error: FMath::FMod(X=-353117930375765884928.000000, Y=0.000000) : Y is zero, this is invalid and would result in NaN!
LogOutputDevice: Error: Stack: 
LogOutputDevice: Error: [Callstack] 0x00007fff45a80410 UnrealEditor-Core.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff30e4a194 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff30c3ddd8 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff30c26d42 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff30c82c33 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff300ea141 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff30c32d04 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff30bff967 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff30c17ff1 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff30c1fe18 UnrealEditor-Renderer.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff459e2fdd UnrealEditor-Core.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff459e352e UnrealEditor-Core.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff4090f804 UnrealEditor-RenderCore.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff40913b54 UnrealEditor-RenderCore.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff46105342 UnrealEditor-Core.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fff460fd230 UnrealEditor-Core.dll!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fffafd626bd KERNEL32.DLL!UnknownFunction []
LogOutputDevice: Error: [Callstack] 0x00007fffb174a9f8 ntdll.dll!UnknownFunction []
```

I hope it helps a bit to find why I got NaNs.

**Comment by seanmcleod** [comment]
> - Get some altitude ~100ft for instance
> - Push the elevator stick (for instance to simulate an engine failure)
>
>The airplane start to fall forward and get speed, then I got a lot of NaN logged in the console with an exception. If the angle is low, I can land without problem. If the angle of attack is to high it gets NaN.

So I don't see any comment about trying to flare/land? Just push the stick forward and 'crash' into the ground?

> If the angle of attack is too high 

Or are you flaring and increasing the AoA, or do you mean the flight path angle?

Assuming more of a 'crash' into the ground I wonder if you aren't running into this issue/discussion - https://github.com/JSBSim-Team/jsbsim/discussions/819

**Reply by demonixis** [reply]
I push the elevator stick forward to makes the nose go down. I'm high enough to not crash the airplane of course. When doing this, if the angle is to high and fast (actually it's negative because I'm going in the direction of the ground), approx. -35° it gets NaNs.

So no it's not a crash at all, I can make a video to show you the problem. In real condition we're rarely in that situation, except when simulating an engine failure during takeoff. But when flying, there are situations where you can get a fast angle change. It crashes in that case too.

**Reply by demonixis** [reply]
Following my reply here is a [video that show you the problem](https://youtu.be/8eBmoKhTlHs)

As you can see, the crash happens just when a release the elevator stick. We can see the values on the debug screen. Elevator values are clamped to -1/1.

**Comment by seanmcleod** [comment]
Okay

So at 51s into the video you have an AGL of 141.54ft and VS of -30ft/s and none of the gear/contact points have WOW (Weight On Wheel) set.

![image](https://user-images.githubusercontent.com/824856/229566400-d118045f-bf12-436e-a0c0-962fa5772d66.png)

And then at 52s into the video you have landing gear/contact point 0 with WOW and an AGL of 2.57ft even though the altimeter shows you > 100ft.

![image](https://user-images.githubusercontent.com/824856/229567382-b28f0a1d-d9ce-4e49-aa86-656e2441dc6e.png)

There is also a video frame in-between showing a WOW while AGL is 137.43ft.

![image](https://user-images.githubusercontent.com/824856/229568050-cbd5c1ea-8a03-412a-be9e-d83f76aa122f.png)

So looks like some terrain lookup going bad?

And it seems it could be similiar to this issue - https://github.com/JSBSim-Team/jsbsim/issues/711 although in that case you could see a massive force value when WOW was triggered.

Note the comment in that issue:

> We use to try with the C172p or c182 aircrafts, ande we have the problem. But when we launched with the 737 (in a Cesium level), we could fly for more than an hour without any physics/stability issue. Maybe the problem is related to the Cessna files ?

**Comment by gallonmate** [comment]
Uncommenting the debug line seen below, in the GetALevel function and also checking the name of the component in the HitResult would confirm if the aircraft is self-colliding or colliding with an unintended object.

https://github.com/JSBSim-Team/jsbsim/blob/10ac2036d8b5ac6d840a105f41062c603e952e6a/UnrealEngine/Plugins/JSBSimFlightDynamicsModel/Source/JSBSimFlightDynamicsModel/Private/JSBSimMovementComponent.cpp#L218-L222

**Comment by demonixis** [comment]
Well I think I found the problem, sometimes the `HitResult` detects my player, however it doesn't have any active collisions so I don't understand why.

I changed the code like this

```cpp
FHitResult HitResult = FHitResult();
static const FName LineTraceSingleName(TEXT("AGLevelLineTrace"));
FCollisionQueryParams CollisionParams(LineTraceSingleName);
CollisionParams.bTraceComplex = true;
CollisionParams.AddIgnoredActor(Parent);

FCollisionObjectQueryParams ObjectParams = FCollisionObjectQueryParams(ECC_WorldStatic);
//ObjectParams.AddObjectTypesToQuery(ECC_WorldDynamic);
//ObjectParams.AddObjectTypesToQuery(ECC_Pawn);
//ObjectParams.AddObjectTypesToQuery(ECC_Visibility);
```

Now it works well all the time without any NaN

Is it fine if I submit a pull request with a parameter to only line trace on static objects instead of static + dynamic?

**Reply by AlbanBERGERET-Epic** [reply]
Great that you found it! 

Actually, I spent 10mn with the reference app doing any sort of acrobatics with the c172 without ever having a Nan or a crash! 

And this collision with your player makes sense. Depending on the angle of your aircraft, the FDM receives incoherent inputs for what he expects to be under him. Not surprising you receive NaNs.

With regard to the appropriate solution, I don't think it's a good idea to change the line traces target. We need to handle all cases, like for instance landing on a moving aircraft carrier. 

Two solutions: 

- Remove/disable the collision on your player actor if you don't need them (or when you board the plane)
- Add an array of IgnoredActors for collision in the JSBSimMovement component, add this on top of the CollisionParams.AddIgnoredActor(Parent), and include your player there.

**Comment by demonixis** [comment]
I think the second solution is just fine. I can create the PR with this change tomorrow. Still doing some tests. Now it works, I'm playing a bit with the physics and it's close to a real aircraft, I love it :)

**Comment by gallonmate** [comment]
There is no need for a pull request. These changes would be useful for your specific project and your game design.

Also for reference, I have no issues when attaching a Character class to the aircraft. I disabled collision on the Capsule Component of the Character and everything works normally with no bugs.
